ada_feeding/ada_feeding/trees/acquire_food_tree.py

```diff
@@ -63,9 +72,67 @@ def create_tree(
                             # Default food_frame_id = "food"
                             # Default world_frame = "world"
                         },
-                        outputs={"action_select_request": None, "food_frame": None},
+                        outputs={
+                            "action_select_request": BlackboardKey("action_request"),
+                            "food_frame": None,
+                        },
+                    ),
+                ),
+                # Get Action to Use
+                py_trees_ros.service_clients.FromBlackboard(
+                    name="AcquisitionSelect",
+                    service_name="~/action_select",
+                    service_type=AcquisitionSelect,
+                    # Need absolute Blackboard name
+                    key_request=Blackboard.separator.join(
+                        [name, BlackboardKey("action_request")]
+                    ),
+                    key_response=Blackboard.separator.join(
+                        [name, BlackboardKey("action_response")]
                     ),
-                )
+                    # Default fail if service is down
+                    wait_for_server_timeout_sec=0.0,
+                ),
+                # Get MoveIt2 Constraints
+                ComputeActionConstraints(
+                    name="ComputeActionConstraints",
+                    ns=name,
+                    inputs={
+                        "action_select_response": BlackboardKey("action_response"),
+                        # Default move_above_dist_m = 0.05
+                    },
+                    outputs={
+                        "move_above_pose": BlackboardKey("move_above_pose"),
+                    },
+                ),
+                # Move Above Food
+                MoveIt2PoseConstraint(
```

Comment:
(1) Perhaps we should consider having a helper function at a higher level that clears the blackboard before each tree run

(2) I can add this for convenience, but note that the behavior takes both an input and an output constraints array, so if the input constraints are unset or None, then it will create a new one to write to the blackboard, solving the multiple runs problem

Reply:
(1) I don’t think we need a higher level function, because most blackboard inputs are reset in initialise. `constraints` is the notable exception, because an initialise call can’t differentiate between the case where it is getting called on a new tree, or the constraint was populated by prior behaviors.
(2) I’m not sure how this solves the multiple runs problem? I agree that if constraints is under it will initialize it. The multiple runs issue is that constraints will already be set on a second run, so it will keep appending to that.
(3) another way to address this could be to give each constraint a name/key, and make the constraint container stored on the blackboards a duct. That way it will override the constraint if it’s name was already there.

Reply:
Discussed externally, by passing in None as a constant to the first Move2PoseConstraints, we solve the multi-run problem